catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswEndpoint.java

```diff
@@ -813,6 +843,179 @@ public void unknownOperation() throws CswException {
+                    .build();
+        }
+        return createAcknowledgment(subscription.getOriginalRequest());
+    }
+
+    @Override
+    @PUT
+    @Path("/subscription/{requestId}")
+    @Consumes({MediaType.TEXT_XML, MediaType.APPLICATION_XML})
+    @Produces({MediaType.WILDCARD})
+    public Response updateRecordsSubscription(@PathParam("requestId") String requestId,
+            GetRecordsType request) throws CswException {
+
+        return createOrUpdateSubscription(request, requestId);
+    }
+
+    @Override
+    @GET
+    @Path("/subscription")
+    @Produces({MediaType.WILDCARD})
+    public Response createRecordsSubscription(@QueryParam("") GetRecordsRequest request)
+            throws CswException {
+        if (request == null) {
+            throw new CswException("GetRecordsSubscription request is null");
+        } else {
+            LOGGER.debug("{} attempting to subscribe.", request.getRequest());
+        }
+        if (StringUtils.isEmpty(request.getVersion())) {
+            request.setVersion(CswConstants.VERSION_2_0_2);
+        } else {
+            validateVersion(request.getVersion());
+        }
+
+        LOGGER.trace("Exiting getRecordsSubscription");
+
+        return createOrUpdateSubscription(request.get202RecordsType(), null);
+    }
+
+    @Override
+    @POST
+    @Path("/subscription")
+    @Consumes({MediaType.TEXT_XML, MediaType.APPLICATION_XML})
+    @Produces({MediaType.WILDCARD})
+    public Response createRecordsSubscription(GetRecordsType request) throws CswException {
+        return createOrUpdateSubscription(request, null);
+    }
+
+    private Response createOrUpdateSubscription(GetRecordsType request, String requestId)
+            throws CswException {
+
+        if (request == null) {
+            throw new CswException("Request is null");
+        } else {
+            LOGGER.debug("{} attempting to subscribe.", request.getService());
+        }
+
+        validateOutputFormat(request.getOutputFormat());
+
+        validateOutputSchema(request.getOutputSchema());
+
+        if (request.getAbstractQuery() != null) {
+            if (!request.getAbstractQuery()
+                    .getValue()
+                    .getClass()
+                    .equals(QueryType.class)) {
+                throw new CswException("Unknown QueryType: " + request.getAbstractQuery()
+                        .getValue()
+                        .getClass());
+            }
+
+            QueryType query = (QueryType) request.getAbstractQuery()
+                    .getValue();
+
+            validateTypes(query.getTypeNames(), CswConstants.VERSION_2_0_2);
+
+            validateElementNames(query);
+
+            if (query.getConstraint() != null &&
+                    query.getConstraint()
+                            .isSetFilter() && query.getConstraint()
+                    .isSetCqlText()) {
+                throw new CswException("A Csw Query can only have a Filter or CQL constraint");
```

Comment:
Both are user error do we typically log it when the exception will generate a nice response with this message in it?